modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java

```diff
@@ -0,0 +1,487 @@
+    private @Nullable SchemaDescriptor searchSchemaByVersion(int tblId, int schemaVer) {
+        SchemaRegistry registry = registriesVv.latest().get(tblId);
+
+        if (registry != null && schemaVer <= registry.lastSchemaVersion()) {
+            return registry.schema(schemaVer);
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Get the schema registry for the given causality token and table id.
+     *
+     * @param causalityToken Causality token.
+     * @param tableId Id of a table which the required registry belongs to. If {@code null}, then this method will return a future which
+     *     will be completed with {@code null} result, but only when the schema manager will have consistent state regarding given causality
+     *     token.
+     * @return A future which will be completed when schema registries for given causality token are ready.
+     */
+    public CompletableFuture<SchemaRegistry> schemaRegistry(long causalityToken, int tableId) {
+        if (!busyLock.enterBusy()) {
+            throw new IgniteException(NODE_STOPPING_ERR, new NodeStoppingException());
+        }
+
+        try {
+            return registriesVv.get(causalityToken)
+                    .thenApply(regs -> inBusyLock(busyLock, () -> regs.get(tableId)));
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    /**
+     * Returns schema registry by table id.
+     *
+     * @param tableId Table id.
+     * @return Schema registry.
+     */
+    public SchemaRegistry schemaRegistry(int tableId) {
+        return registriesVv.latest().get(tableId);
+    }
+
+    /**
+     * Drop schema registry for the given table id.
+     *
+     * @param causalityToken Causality token.
+     * @param tableId Table id.
+     */
+    public CompletableFuture<?> dropRegistry(long causalityToken, int tableId) {
+        return registriesVv.update(causalityToken, (registries, e) -> inBusyLock(busyLock, () -> {
+            if (e != null) {
+                return failedFuture(new IgniteInternalException(
+                        IgniteStringFormatter.format("Cannot remove a schema registry for the table [tblId={}]", tableId), e));
+            }
+
+            Map<Integer, SchemaRegistryImpl> regs = new HashMap<>(registries);
+
+            regs.remove(tableId);
+
+            return completedFuture(regs);
+        }));
+    }
+
+    @Override
+    public void stop() throws Exception {
+        if (!stopGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+    }
+
+    /**
+     * Gets the latest version of the table schema which is available in Metastore.
+     *
+     * @param tblId Table id.
+     * @return The latest schema version.
+     */
+    private CompletableFuture<Integer> latestSchemaVersion(int tblId) {
+        var latestVersionFuture = new CompletableFuture<Integer>();
+
+        metastorageMgr.prefix(schemaHistPrefix(tblId)).subscribe(new Subscriber<>() {
```

Comment:
I think we can update 2 keys in single transaction
* write a new descriptor to the history
* update latest version key with a version from written descriptor.